src/Validator/ValidAddProductBundleToCartCommandValidator.php

```diff
@@ -0,0 +1,157 @@
+     * @param AddProductBundleToCartCommand|mixed $value
+     * @param ValidAddProductBundleToCartCommand|Constraint $constraint
+     */
+    public function validate($value, Constraint $constraint): void
+    {
+        Assert::isInstanceOf($value, AddProductBundleToCartCommand::class);
+        Assert::isInstanceOf($constraint, ValidAddProductBundleToCartCommand::class);
+
+        if (null === $value->getOrderId() && null === $value->getOrderToken()) {
+            $this->context->addViolation(
+                $constraint::NO_ORDER_ID_OR_TOKEN_MESSAGE
+            );
+
+            return;
+        }
+
+        /** @var ProductBundleInterface|null $productBundle */
+        $productBundle = $this->productBundleRepository->find($value->getProductBundleId());
+
+        if (null === $productBundle) {
+            $this->context->addViolation(
+                $constraint::PRODUCT_BUNDLE_DOESNT_EXIST_MESSAGE,
+                [
+                    '{{ id }}' => $value->getProductBundleId(),
+                ]
+            );
+
+            return;
+        }
+
+        $product = $productBundle->getProduct();
+        if (!$product->isEnabled()) {
+            $this->context->addViolation(
+                $constraint::PRODUCT_DISABLED_MESSAGE,
+                [
+                    '{{ code }}' => $product->getCode(),
+                ]
+            );
+
+            return;
+        }
+
+        /** @var ProductVariantInterface $productVariant */
+        $productVariant = $product->getVariants()->first();
+        if (!$productVariant->isEnabled()) {
+            $this->context->addViolation(
+                $constraint::PRODUCT_VARIANT_DISABLED_MESSAGE,
+                [
+                    '{{ code }}' => $productVariant->getCode(),
+                ]
+            );
+
+            return;
+        }
+
+        /** @var \Sylius\Component\Core\Model\OrderInterface|null $cart */
+        $cart = $this->getCart($value);
+        if (null === $cart) {
+            $this->context->addViolation(
+                $constraint::CART_DOESNT_EXIST_MESSAGE
+            );
+
+            return;
+        }
+
+        $targetQuantity = $value->getQuantity() + $this->getCurrentProductVariantQuantityFromCart($cart, $productVariant);
+        if (!$this->availabilityChecker->isStockSufficient($productVariant, $targetQuantity)) {
+            $this->context->addViolation(
+                $constraint::PRODUCT_VARIANT_INSUFFICIENT_STOCK_MESSAGE,
+                [
+                    '{{ code }}' => $productVariant->getCode(),
+                ]
+            );
+
+            return;
+        }
+
+        $channel = $cart->getChannel();
+        Assert::notNull($channel);
+
+        if (!$product->hasChannel($channel)) {
+            $this->context->addViolation(
+                $constraint::PRODUCT_DOESNT_EXIST_MESSAGE,
+                [
+                    '{{ name }}' => $product->getName(),
+                ]
+            );
+        }
+    }
```

Comment:
Could this method be shortened?

Reply:
In terms of logic I'm afraid not. The only thing I can do with this method's length is splitting "stages" into private methods. Would it be better or left it as is?